### ISSUE TITLE
native.sh_binary converts binary arguments starting with /

### ISSUE THREAD

**Issue by filipesilva** [bazelbuild/bazel#6764]
### Description of the problem / feature request:

When using Bazel under Windows and using `native.sh_binary` with a script that invokes a binary, arguments starting with `/` will be converted to paths according to [Posix_path_conversion]( http://www.mingw.org/wiki/Posix_path_conversion).

This becomes a problem when constructing scripts that need to take user input for which this format might be relevant, such as specifying a serving path for a http server (https://github.com/bazelbuild/rules_typescript/pull/327#discussion_r236315538).

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Have a script that invokes a binary, providing it with an argument starting with `/`.
```
echo /something
node -e "console.log(process.argv)" /something
```
```
native.sh_binary(
    name = "script",
    srcs = ["script.sh"],
)
```

I used `node` as the binary because I had it in my system and this didn't seem to manifest with built-ins like `echo`.

Below is a repository with this code.

```
git clone https://github.com/filipesilva/bazel-msys-script
cd bazel-msys-script
bazel run :script
```

You should see
```
INFO: Analysed target //:script (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:script up-to-date:
  C:/users/kamik/_bazel_kamik/f23qfmhz/execroot/__main__/bazel-out/x64_windows-fastbuild/bin/script
  C:/users/kamik/_bazel_kamik/f23qfmhz/execroot/__main__/bazel-out/x64_windows-fastbuild/bin/script.exe
INFO: Elapsed time: 0.498s, Critical Path: 0.01s, Remote (0.00% of the time): [queue: 0.00%, setup: 0.00%, process: 0.00%]
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Running command line: C:/users/kamik/_bazel_kamik/f23qfmhz/execroot/__main__/bazel-out/x64_windows-fastbuild/bin/sINFO: Build completed successfully, 1 total action
echo /something
[ 'C:\\Program Files\\nodejs\\node.exe', 'C:/msys64/something' ]
```

### What operating system are you running Bazel on?

Windows 10 1803.

### What's the output of `bazel info release`?

`0.19.2`.

###  Have you found anything relevant by searching the web?

No.

### Any other information, logs, or outputs that you want to share?

Path conversion behaviour in MSYS can be disabled using `MSYS2_ARG_CONV_EXCL="*"` (https://github.com/Alexpux/MSYS2-packages/issues/84#issuecomment-58821088).

Related to https://github.com/bazelbuild/bazel/issues/3001, https://github.com/bazelbuild/bazel/issues/5966.

**Comment by laszlocsomor** [comment]
Thanks for the report! Do you see any problem with `MSYS2_ARG_CONV_EXCL="*"`? Looks like that would solve the problem.

**Comment by filipesilva** [comment]
@laszlocsomor prepending `MSYS2_ARG_CONV_EXCL` to the binary call (`MSYS2_ARG_CONV_EXCL="*" node -e "console.log(process.argv)" /something`) will disable the path conversion.

But that puts the onus on Bazel developers to be aware of an implementation detail (usage of MSYS on Windows platforms) inside of Bazel. I'm not sure how many are aware of this caveat.

Are there any cases where the path conversion is desirable? Perhaps the flag could be set by Bazel when invoking MSYS.

Off the top of my head:
- inside shell scripts proper one is always using win32
- any path sent in by the user running Bazel on a MSYS shell will reach Bazel already converted

**Comment by laszlocsomor** [comment]
Here's an even better demo: https://github.com/laszlocsomor/projects/tree/c3caa558c7fe59388df734b9f59baac92dae86fe/bazel/msys-path-conv

**Comment by filipesilva** [comment]
Well there are three types of developers, so it is worth to clarify:
- t1: Bazel developers (e.g. developers on https://github.com/bazelbuild/bazel)
- t2: Bazel rules developers (e.g. developers on https://github.com/bazelbuild/rules_nodejs)
- t3: Developers/users of Bazel (e.g. developers that use https://github.com/alexeagle/angular-bazel-example)

When I said "But that puts the onus on Bazel developers to be aware of an implementation detail" I should have clarified. I meant the t2 developers, that have to understand that Bazel uses MSYS2 on windows and safeguard their code against path conversion where applicable.

If in the future MSYS2 is no longer used, t2 developers would to remove MSYS2 logic from their rules. I won't do anything if it stays there but it's essentially dead code.

If t2 developers properly safeguard their rules, t3 developers won't have to care about MSYS2. But if they don't, path conversion can give rise to odd bugs.

I think it's accurate that t1 developers cannot make the choice to blanket disable path conversion logic, given the myriad use cases.

Perhaps it is a matter of documentation for t2 developers on how to properly support Windows. 

I'm not aware of a definite resource for this topic right now, but if I had to choose a few topics they would be:
- wrapping executable calls in native.sh_binary scripts
- using the Bash runfiles library
- creating runfiles helpers for language rules
- dynamic resolution using runfiles
- guarding against MSYS2 path conversion

**Comment by laszlocsomor** [comment]
Thanks for the clarification!

I see no way to relieve t2 developers from having to know about Bazel's use of MSYS2 and about MSYS2's quirks in particular. I agree documentation could help explaining this so t2 developers don't have to reinvent the wheel and learn about these quirks every time.

Where to you think such documentation should live so it's easily discoverable? (Where would you look for it?)

Thanks for the topics! That's a great set Qs for a FAQ.

**Comment by filipesilva** [comment]
I would look for it in https://bazel.build/ -> `Documentation` -> `Rules` -> `Developing your own rules` (a new section.

I'd expect to find a small tutorial on how to make my own language rule and how to ensure windows compatibility, and some resources for further reading.

**Comment by laszlocsomor** [comment]
Thanks! I think https://github.com/bazelbuild/bazel/issues/3889 is the most relevant bug.

**Comment by laszlocsomor** [comment]
> I would look for it in https://bazel.build/ -> `Documentation` -> `Rules` -> `Developing your own rules` (a new section.
> 
> I'd expect to find a small tutorial on how to make my own language rule and how to ensure windows compatibility, and some resources for further reading.

**Comment by meteorcloudy** [comment]
This issue looks resolved by #3889
